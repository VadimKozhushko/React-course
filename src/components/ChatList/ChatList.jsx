
import { NavLink } from 'react-router-dom';
import {
  Box,
  ListSubheader,
  List,
  ListItemButton,
  Typography } from '@mui/material';
import { useState } from "react"
import { useDispatch, useSelector } from 'react-redux'
import { addChat, deleteChat } from '../../store/messages/actions'
import { selectChat } from '../../store/messages/selectors'
import { push, set, remove } from "firebase/database";
import { messagesRef, getChatById, getMessageListById } from '../../services/firebase';



export function ChatList({messagesDB, chats}) {
    const [value, setValue] = useState('')
    const dispatch = useDispatch()
  
    console.log('update chats', chats)
  
    const handleSubmit = (e) => {
      e.preventDefault()
      dispatch(addChat(value))
  
      set(messagesRef, {
        ...messagesDB,
        [value]: {
          name: value
        }
      })
  
      push(getMessageListById(value), {
        text: 'Chat has been created',
        author: 'Admin',
      });
  
      setValue('');
    }
  
    console.log('chats', chats)
    const handleDeleteChat = (chatId) => {
      remove(getChatById(chatId));
    };
  
    
  return (
    <>
    <Box
    sx={{
        display: 'flex',
        flexDirection: 'column',
        width: '100%',
        maxWidth: '250px',
        bgcolor: 'green',
        borderRadius: '20px',
    }}
>
    <List
        sx={{
            width: '100%',
            borderRadius: '20px',
        }}
        component="nav"
        aria-labelledby="nested-list-subheader"
        subheader={
            <ListSubheader
                component="div"
                id="nested-list-subheader"
                sx={{
                    bgcolor: 'green',
                    color: 'white',
                    borderRadius: '20px',
                }}
            >
                Chats
            </ListSubheader>
        }
    >
        {
            chats.map(chat => {
                return (
                  <ListItemButton
                                component={NavLink}
                                to={`/chats/${chat.name}`}
                                key={chat.id}
                            
                                sx={{
                                    borderBottom: '3px solid transparent',
                                    '&.active': {
                                        backgroundColor: '#0d4b1a',
                                        borderBottom: '3px solid white'
                                    }
                                }}
                            >
                                
                                <Typography
                                    sx={{
                                        color: 'primary.contrastText',
                                    }}
                                >
                                    {chat.name}
                                </Typography>
                                <button onClick={() => dispatch(handleDeleteChat(chat.name))}>X</button>
                            </ListItemButton>
                )
            })
        }
    </List>
</Box>
<h1>ChatList</h1>
      <form onSubmit={handleSubmit}>
        <input 
          type="text" 
          value={value}
          onChange={(e) => setValue(e.target.value)}
        />
        <button type="submit">Create Chat</button>
      </form>

</>
)


}
