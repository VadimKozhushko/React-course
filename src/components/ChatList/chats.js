import { nanoid } from 'nanoid'

const chats = [
  {
      id: nanoid(),
      name: 'chat1'
      
  },
  {
      id: nanoid(),
      name: 'chat2'
      
  },
  {
      id: nanoid(),
      name: 'chat3'
      
  },
  {
      id: nanoid(),
      name: 'chat4'
      
  }
]
export default chats