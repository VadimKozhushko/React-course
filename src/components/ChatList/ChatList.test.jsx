import React from "react";
import { render, screen } from "@testing-library/react";
import '@testing-library/jest-dom'

import { ChatList } from "./ChatList";

import * as reduxHooks from 'react-redux'


jest.mock('react-redux')
const mockUseDispatch = jest.spyOn(reduxHooks, 'useDispatch')
describe('testing ChatList', () => {
  it('should create ChatList', () => {
    mockUseDispatch.mockReturnValue(jest.fn)
    const component = render(<ChatList messageDB={{}} chats={[]} />)
    expect(component).toMatchSnapshot()
  })

  it('should dispatch action', () => {
    const dispatch = jest.fn;
    mockUseDispatch.mockReturnValue(dispatch)
   
    render(
      <ChatList messageDB={{}} chats={[]} />
      )
    const buttons = screen.getByRole('button')

  })

})