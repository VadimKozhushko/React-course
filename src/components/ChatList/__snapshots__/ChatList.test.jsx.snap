// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ChatList should create ChatList 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-10vrplw"
      >
        <nav
          aria-labelledby="nested-list-subheader"
          class="MuiList-root MuiList-padding MuiList-subheader css-nj6nf1-MuiList-root"
        >
          <div
            class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-1sql39t-MuiListSubheader-root"
            id="nested-list-subheader"
          >
            Chats
          </div>
        </nav>
      </div>
      <h1>
        ChatList
      </h1>
      <form>
        <input
          type="text"
          value=""
        />
        <button
          type="submit"
        >
          Create Chat
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-10vrplw"
    >
      <nav
        aria-labelledby="nested-list-subheader"
        class="MuiList-root MuiList-padding MuiList-subheader css-nj6nf1-MuiList-root"
      >
        <div
          class="MuiListSubheader-root MuiListSubheader-gutters MuiListSubheader-sticky css-1sql39t-MuiListSubheader-root"
          id="nested-list-subheader"
        >
          Chats
        </div>
      </nav>
    </div>
    <h1>
      ChatList
    </h1>
    <form>
      <input
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Create Chat
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
